rfcs/003-double-entry-ledger.md

```diff
@@ -0,0 +1,122 @@
+# REQUEST FOR COMMENTS 003 - Double Entry Ledger
+
+
+## Overview
+
+Transferring funds from User to Collective is currently represented as
+two rows in the `Transactions` table. One to add up to the Collective
+ledger (CREDIT) and the other to subtract from the User's ledger
+(DEBIT).
+
+However the transactions between the collective and the host, the
+platform and the payment processor (fees) don't have their own
+separate rows. The fees are stored as fields in both database rows but
+that don't quite describe the flow of funds between the collective and
+the institutions that are receiving the fees.
+
+This change will increase accuracy because it will simplify the way we
+store the transactions. It will also decrease duplicated code (likely
+reducing bugs) because it will provide a library that implements all
+the basic operations related to the ledger.
+
+## Goals
+
+1. Increase accuracy of the ledger
+   1. [ ] Create collective type "Payment Method"
+   2. [ ] Create separate transactions for fees
+2. Make it easier to build new features that depend on data from the
+   ledger
+   1. [ ] `libocledger`: wrap the most common operations related to
+           the ledger under a library that can be used across all the
+           places that interact with the ledger and decrease
+           duplicated code in the repository.
+
+### Create collective type "Payment Method"
+
+As mentioned before, the amount of the payment processor fee is stored
+in the field `paymentProcessorFeeInHostCurrency` in the `Transactions`
+table. To find out to which specific payment method the fee went to,
+the `Transactions` table also records the `PaymentMethodId`.
+
+This is different from how transactions are tracked between User &
+Collective. To address this difference, the Payment Method needs to
+have a ledger as well.
+
+### Create separate transactions for fees
+
+#### Use case example
+
+Given a transaction that represents moving $5 dollars from User to
+Collective with 5% of platform fee, 10% of host fee and 2.9%+30c of
+payment processor fee.
+
+##### How it is currently stored
+|            | User Ledger | Collective Ledger |
+|------------|------------:|------------------:|
+| Type       |       DEBIT |            CREDIT |
+| Amount     |       -4075 |              5000 |
+| Host Fee   |        -500 |              -500 |
+| Plat Fee   |        -250 |              -250 |
+| PP Fee     |        -175 |              -175 |
+| Net Amount |       -5000 |              4075 |
+
+##### How it should look like after this change
+
+|        |  User | Collective | Collective | Platform | Collective | Host | Collective | Payment Method |
```

Comment:
Might be helpful to represent each of these as rows with "fromCollective" and "toCollective". And then showing the net for each entity at the end of it.

Reply:
Do you think that's an explanation that should go into the documentation or you're talking about the data that will be saved in the table? If it's related to the data in the table, I'd actually answer with another question. Is it useful to separately list all the fields that we'd keep in that table?

Reply:
Not worth listing table fields here. More the idea of double entry. It wasn't clear to me that, for example, `Collective` columns above should be summed to get the netAmount for that Collective.